### DISCUSSION TITLE
Sort terms in metric_aggregation rule with query_key

### DISCUSSION THREAD

**Discussion by baphael** [jertel/elastalert2#844] [Q&A] [answered]
Hello ! 👋 

I have been struggling lately with one specific rule (a metric aggregation of type sum). Below is what my rule looks like : 

```
name: my-rule-agg-sum
type: metric_aggregation
index: my-index*
use_strftime_index: False
doc_type: _doc

metric_agg_key: field1
metric_agg_type: sum
max_threshold: 1000000
buffer_time:
 minutes: 5

filter:
  [...]

alert:
- "email"
email:
- "john.doe@mail.xyz"

query_key:
 - field4

query_delay:
 minutes: 5

realert:
 hours: 1
```

I know for sure that many `field4` (query_key) entities exceed the `max_threshold` value of 1000000. I've tested it by adding a filter for one specific value of `field4` which I knew would've matched and... it matched. That said, without any filter on `field4`, the rule does not send any alert. (For information, I have more than 150000 distinct values of `field4`).

I read in the documentation that by default, elastalert returns 50 terms (can be changed through `terms_size`...). That said I thought that those terms would be sorted by `mertric_field1_sum`... Apparently not. Here is a sample of the result returned by elastic (by using `--es_debug` and `--es-debug-trace <output>` flags) : 

QUERY : 
```
POST /my-index*/_search?pretty&ignore_unavailable=true&size=0
{
  "aggs": {
    "bucket_aggs": {
      "aggs": {
        "metric_field1_sum": {
          "sum": {
            "field": "field1"
          }
        }
      },
      "terms": {
        "field": "src.ip",
        "min_doc_count": 1,
        "size": 50
      }
    }
  },
  "query": {
    "bool": {
      "filter": {
        "bool": {
          "must": [
            {
              "range": {
                "@timestamp": {
                  "gt": "2022-05-13T08:16:29.957532Z",
                  "lte": "2022-05-13T08:21:29.957532Z"
                }
              }
            },
            [...]
          ]
        }
      }
    }
  }
}
```

RESPONSE : 
```
{
  "took" : 5896,
  "timed_out" : false,
  "_shards" : {
    "total" : 186,
    "successful" : 186,
    "skipped" : 184,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 6733,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "bucket_aggs" : {
      "doc_count_error_upper_bound" : 10,
      "sum_other_doc_count" : 5821,
      "buckets" : [
        {
          "key" : "field4value1",
          "doc_count" : 63,
          "metric_field1_sum" : {
            "value" : 126.0
          }
        },
        {
          "key" : "field4value2",
          "doc_count" : 54,
          "metric_field1_sum" : {
            "value" : 166.0
          }
        },
        {
          "key" : "field4value3",
          "doc_count" : 42,
          "metric_field1_sum" : {
            "value" : 105.0
          }
        },
        {
          "key" : "field4value4",
          "doc_count" : 25,
          "metric_field1_sum" : {
            "value" : 130.0
          }
        },
        {
          "key" : "field4value5",
          "doc_count" : 23,
          "metric_field1_sum" : {
            "value" : 55.0
          }
        },
		[...]
        {
          "key" : "field4value50",
          "doc_count" : 10,
          "metric_field1_sum" : {
            "value" : 150.0
          }
        }
      ]
    }
  }
}
```

So, my question are :
- Is there a way to force ElastAlert to sort terms by metric_agg_key values in an aggregation rule ?
- Am I missing something stupid/simple ?
- Does this behaviour has anything to do with `"doc_count_error_upper_bound" : 10,` in the ES response ?

Sorry if this somehow turns out to be a duplicate, and many thanks in advance for your time ! 🙂

**Comment by jertel** [comment, marked as answer]
I am not aware of a way to explicitly order aggregations in ElastAlert 2. By default, it will order the aggregation top terms by document _count_, not by the aggregation computation. I believe it could be done but would require someone to implement it via a pull request. Essentially, this block of code in `elastalert.py` that constructs the agg query:

```
                aggs_element = {'bucket_aggs': {'terms': {'field': key, 'size': terms_size,
                                                          'min_doc_count': rule.get('min_doc_count', 1)},
                                                'aggs': aggs_element}}
```
Needs to be modified to support an `order` key/value pair, using the dynamically generated agg name based on the aggregated field name + computation method. Similar to the following:
```
            "order": { "metric_field1_sum": "desc" }
```
That said, I don't have much experience in this area as it relates to Elastic shards and how each elastic search node computes its top terms and then how the coordinator node computes the final top terms, and whether this `order` instruction will be respected across the entire search flow. More information on the Elastic agg order behavior is located here: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#search-aggregations-bucket-terms-aggregation-order

The other option is to see if increasing `term_size` will solve your problem, by forcing each node to return a larger number of top terms, instead of the default 50. The larger this value is, the more memory your search nodes will need. Place inside the rule yaml file:

```
term_size: 500
```
